src/main/java/com/thealgorithms/strings/AhoCorasick.java

```diff
@@ -0,0 +1,228 @@
+ *                                                                               *
+ * This code implements the Aho-Corasick algorithm, which is used for efficient  *
+ * string matching in a given text. It can find multiple patterns simultaneously *
+ *                 and records their positions in the text.                      *
+ *                                                                               *
+ *                      Author: Prabhat-Kumar-42                                 *
+ *               GitHub: https://github.com/Prabhat-Kumar-42                     *
+ *                                                                               *
+ *********************************************************************************/
+
+package com.thealgorithms.strings;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.Map;
+import java.util.Queue;
+
+public class AhoCorasick {
+
+    // Trie Node Class
+    public class Node {
+        // Represents a character in the trie
+        private HashMap<Character, Node> child = new HashMap<>(); // Child nodes of the current node
+        private Node suffixLink; // Suffix link to another node in the trie
+        private Node outputLink; // Output link to another node in the trie
+        private int patternInd; // Index of the pattern that ends at this node
+
+        Node() {
+            this.suffixLink = null;
+            this.outputLink = null;
+            this.patternInd = -1;
+        }
+
+        public HashMap<Character, Node> getChild() {
+            return child;
+        }
+
+        public Node getSuffixLink() {
+            return suffixLink;
+        }
+
+        public void setSuffixLink(Node suffixLink) {
+            this.suffixLink = suffixLink;
+        }
+
+        public Node getOutputLink() {
+            return outputLink;
+        }
+
+        public void setOutputLink(Node outputLink) {
+            this.outputLink = outputLink;
+        }
+
+        public int getPatternInd() {
+            return patternInd;
+        }
+
+        public void setPatternInd(int patternInd) {
+            this.patternInd = patternInd;
+        }
+    }
+
+    // Trie Class
+    public class Trie {
+
+        private Node root = null; // Root node of the trie
+
+        /*
+         * For simplification of methods, declared the below as class attributes
+         *
+         * res is the list of list containing the indexes of words in patterns/dictionary
+         * list index represents word.
+         * eg - list[0] contains the list of start-index of word pattern[0]
+         *      list[1] contains the list of start-index of word pattern[1]
+         *      ......
+         *      ......
+         *      list[n] contains the list of start-index of word pattern[n]
+         */
+
+        private ArrayList<ArrayList<Integer>> res; // Stores positions where patterns are found in the text

```

Comment:
Why is this a member? I think it should be a local variable in `searchIn`.

Reply:
I made it a member because : 

   - I need it to allocate list of all the patterns[] members positions and to do that I have to fist assign it's size as patterns[] which I did in constructor, but I also have to assign all it's i-th position i.e. res[i] = empty ArrayList<>(). If i do all that in patterns I have to I have to run the loop ( which is equal to patterns.length ) two times (2n times if n is size of patterns)

       - 1st time for building `Trie`

       -  2nd time for this empty ArrayList<>() allocation to `res` indices in `searchIn`.

  - If I let it be member : 

       - I only have to run the loop for a single time, i.e. when building the tree in constructor the object, empty ArrayList<>() will be assigned to all res indices simultaneously.

       - I don't have to pass lengthy parameter to setUpStartPoints().

***TL;DR : it was just to save time by running the loop a single time instead of twice.***

Although Since you mentioned I should make it local variable in 'searchIn' . I will implement that, If there's any Issue, please do tell me, I will address it asap.